File: leetcode-js/567.js
/**
 * @param {string} s1
 * @param {string} s2
 * @return {boolean}
 */
var checkInclusion = function (s1, s2) {
    if (s1.length > s2.length) return false;

    let balanceMap = new Map();
    Array.from(s1).forEach(char => balanceMap.set(char, (balanceMap.get(char) || 0) + 1));
    let uniqueLettersCnt = Array.from(balanceMap.keys()).length;

    let left = 0;
    let right = s1.length - 1;
    let balanceSum = 0;
    Array.from(s1).forEach((_, idx) => {
        if (balanceMap.has(s2[idx])) {
            balanceMap.set(s2[idx], balanceMap.get(s2[idx]) - 1);
            if (balanceMap.get(s2[idx]) === 0) {
                balanceSum += 1;
            }
        }
    });
    // console.log(balanceSum, balanceMap);

    while (right < s2.length - 1 && balanceSum !== uniqueLettersCnt) {
        left++;
        right++;
        if (balanceMap.has(s2[left - 1])) {
            if (balanceMap.get(s2[left - 1]) === 0) {
                balanceSum -= 1;
            }
            balanceMap.set(s2[left - 1], balanceMap.get(s2[left - 1]) + 1);
        }

        if (balanceMap.has(s2[right])) {
            balanceMap.set(s2[right], balanceMap.get(s2[right]) - 1);
            if (balanceMap.get(s2[right]) === 0) {
                balanceSum += 1;
            }
        }
        // console.log(balanceMap, balanceSum, s2.slice(left, right + 1), left, right);
    }

    return balanceSum === uniqueLettersCnt;
};

let s1 = "adc";
let s2 = "dcda";
console.log(checkInclusion(s1, s2));

File: leetcode-go/567.go
package main

func checkInclusion(s1 string, s2 string) bool {
	if len(s1) > len(s2) {
		return false
	}
	letterTrackMap := map[rune]int{}

	for _, char := range s1 {
		letterTrackMap[char] = letterTrackMap[char] + 1
	}

	cntLettersCompared := 0
	left := 0
	right := 0
	s2Runes := []rune(s2)
	for right < len(s1) {
		if letterCnt, ok := letterTrackMap[s2Runes[right]]; ok {
			if letterCnt == 0 {
				cntLettersCompared--
			} else if letterCnt-1 == 0 {
				cntLettersCompared++
			}
			letterTrackMap[s2Runes[right]]--
		}
		right++
	}
	right--

	for {
		if cntLettersCompared == len(letterTrackMap) {
			return true
		}
		if letterCnt, ok := letterTrackMap[s2Runes[left]]; ok {
			if letterCnt == 0 {
				cntLettersCompared--
			} else if letterCnt+1 == 0 {
				cntLettersCompared++
			}
			letterTrackMap[s2Runes[left]]++
		}
		left++
		right++
		if right >= len(s2) {
			break
		}
		if letterCnt, ok := letterTrackMap[s2Runes[right]]; ok {
			if letterCnt == 0 {
				cntLettersCompared--
			} else if letterCnt-1 == 0 {
				cntLettersCompared++
			}
			letterTrackMap[s2Runes[right]]--
		}
	}

	return false
}

// func main() {
// 	s1 := "adc"
// 	s2 := "dcda"
// 	fmt.Println(checkInclusion(s1, s2))
// }
